### PR TITLE
[Feature] Block user [Fix] current chat in bold on Linux [Fix] Username is bold on Linux

### LINKED ISSUE #1660: Add functionality to Block a user
[comment]: # (Please replace ... with your information. Remove < and >)

### User Story

As a user I want to block another app user who knows my public key, so even if this user sends me a message I do not see it. 

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Feature

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
*Summary*: It's possible to delete a contact or chat but it does not stop deleted contact from sending me new messages because my public key is known to that user. I'd like to have possibility to block this user, so no new messages of any type are shown to me in the app.

### LINKED ISSUE #6033: Remove logic from desktop messages list view
### User Story

As a developer, I want to remove logic from view

### Description

*Type*: Improvement

*Summary*: Discussed with @yenda that having logic inside view in `src/status_im/ui/screens/desktop/main/chat/views.cljs` isn't ok. At the moment of creating PR #6006 it wasn't clear how to remove it, so this issue created for later fix.

#### Expected behavior
View doesn't contain logic

#### Actual behavior
View contains logic

### Reproduction
- Open `src/status_im/ui/screens/desktop/main/chat/views.cljs`
- check `messages-view` implementation

### Solution
[comment]: # (Please summarise the solution and provide a task list on what needs to be fixed.)
*Summary*:  TBD

### Additional Information
[comment]: # (Please do your best to fill this out.)
* Status version: V0.0.1 (ed2abf910)
[comment]: # (Android or iOS.)
* Operating System: Mac

### PR DESCRIPTION
**Pull request by yenda**
fix #6033 remove logic from view
fix #1660 block user

- usernames in bold font on Linux
- prevent use of rtl char in username to impersonate
other other user random name
- current chat in bold font
- remove defview and subscriptions to use pure functions with params
- allow user to block/unblock a contact
- this will remove blocked user from chat list and
hid any message sent on any channel by blocked user

# Step to test

Plateform: desktop

sanity check
- test that chats, public chats and private chats are working correctly, responses are displayed properly
- test that users can be blocked in their profile

status: ready

### CODE REVIEW

**Comment by flexsurfer** [comment]
@yenda could you please provide some description?

**Review by janherich — APPROVED** [review]

**Comment by dmitryn** [comment]
@yenda please rebase and squash commits

**Comment by yenda** [comment]
@flexsurfer done
@dmitryn done

File: src/status_im/ui/screens/desktop/main/chat/views.cljs
Comment (dmitryn) [inline src/status_im/ui/screens/desktop/main/chat/views.cljs]: shouldn't  `network-status` be reagent subscription? now it looks like as const vector `[:network-status]`

**Review by dmitryn — CHANGES_REQUESTED** [review]

**Review by dmitryn — APPROVED** [review]

**Review by flexsurfer — APPROVED** [review]

**Comment by churik** [comment]
@yenda 
New issue appeared, reproducible only in PR build.
Chat view is broken if you send several messages to chat with last message from yesterday or if you set custom date in the past in your settings (whole or part of chat area is blank)

Possible to reproduce if you have chats with last messages from yesterday or if you set custom date on your laptop to the past.

To emulate it from scratch:
- set your Mac OSx date to past (i.e. yesterday)
- create public channel
- try to send message to this channel

Expected: can see message
Actual: can see only blank screen, no new messages are shown in chat area
Video: http://take.ms/0bpgs
Logs: [Status.log](https://github.com/status-im/status-react/files/2565736/Status.log), [geth.log](https://github.com/status-im/status-react/files/2565738/geth.log)


<blockquote><div>Monosnap screenshot tool</div><div><strong><a href="https://monosnap.com/file/HhZzlXt9KBJmjtYUI8TWyBhq2wsD1W">File "screencast 2018-11-09 12-57-54.mp4"</a></strong></div><div>Monosnap — the best tool to take & share your screenshots.</div></blockquote>

**Comment by yenda** [comment]
@churik I fixed it, I think an easiest way to reproduce was to open a new public chat and then switch to already open ones, at least for me it was like this.

This PR now introduces the following glitch when replying to a chat:

screenshot from 2018-11-09 15-44-47

Personally I don't think this is blocking, and it should make fixing flat-list a priority wdyt ?

**Comment by churik** [comment]
@yenda 
I believe it should be fixed before merge.
Thanks for creating corresponding issue on RN-side!
Blocked by https://github.com/status-im/react-native-desktop/issues/391

File: src/status_im/chat/subs.cljs
Comment (cammellos) [inline src/status_im/chat/subs.cljs]: Please add some tests covering this (at least the newly written code).

File: src/status_im/contact/subs.cljs
Comment (cammellos) [inline src/status_im/contact/subs.cljs]: extract and test?

**Review by cammellos — CHANGES_REQUESTED** [review]
Please add some tests to cover new functionalities

File: src/status_im/chat/subs.cljs
Comment (yenda) [reply src/status_im/chat/subs.cljs]: I was planning to come back to this whole file for a cleanup when doing the feature for mobile but since it's blocked by this flatlist bug I might include the mobile as well

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (dmitryn) [inline src/status_im/ui/screens/chat/message/message.cljs]: looks like too much indent here

**Review by dmitryn — APPROVED** [review]

File: src/status_im/ui/screens/desktop/main/chat/views.cljs
Comment (cammellos) [inline src/status_im/ui/screens/desktop/main/chat/views.cljs]: translations

**Review by cammellos — APPROVED** [review]
Too big to really understand what has changed or follow the logic, please smaller PRs :) I think for plain refactoring small prs with `skip-qa` is ok , I went line by line to see if there's anything obviously odd, but looks ok to me. Please test as well pairing & group chats on a rebased version. 
5000 SNT on my nameday if no tests :)

**Comment by yenda** [comment]
@camellos yes 5000SNT by Nov 30th if I didn't add new tests by then
Regarding the length it's already cut into smaller commits, the issue was that the first commit on which it is built was blocked by a UI glitch in react-native-desktop

**Comment by cammellos** [comment]
@yenda my bad, didn't see that, thanks for pointing out

**Comment by churik** [comment]
Testing is still WIP, comment will be updated:

Mac OSx:
###  1. Last row is cut in chat list if message contains 1+ rows*
Screen:
macos

### 2. Space between last message and previous ones are different
hsdf
The space between last message and text input area is also depending on message is incoming or outgoing:
outg
inc

### 3. No "Add to contacts" in 1-1 chat when someone adds you as contact
Steps:
1) log in to desktop app as user A
2) tap on Share profile code
3) log in as another user on mobile
4) start new chat > scan QR
5) send message to user A

**Expected result: **

User A can see "Add to contacts " in 1-1 chat

**Actual result: **

add_to_contacts

### 4. Spam window doesn't block user from sending new messages 
Steps:
1) log in to desktop app as user A
2) join any public chat
3) send quickly 5+ messages in a row

**Expected result: **

Can send only 5 messages, then popup appears and user cannot send more until popup will be closed

**Actual result: **
Spam window appears, but user still can send messages 
spam


### 5. Switching beween chats is slowed down 
Video from current PR:  https://monosnap.com/file/4esSAd79M1NTzRKJy6qdGwhZHXCqvs
Video from nightly: http://take.ms/HWVui

<blockquote><div>Monosnap screenshot tool</div><div><strong><a href="https://monosnap.com/file/4esSAd79M1NTzRKJy6qdGwhZHXCqvs">File "screencast 2018-11-14 18-12-05.mp4"</a></strong></div><div>Monosnap — the best tool to take & share your screenshots.</div></blockquote>
<blockquote><div>Monosnap screenshot tool</div><div><strong><a href="https://monosnap.com/file/MakgaGhY8yHzxv7QRZqVtKsFd1RQve">File "screencast 2018-11-14 18-21-50.mp4"</a></strong></div><div>Monosnap — the best tool to take & share your screenshots.</div></blockquote>

**Comment by yenda** [comment]
1. seems like an OSX specific problem, I increased the max size because on Linux the bottom of some letters was cut. I changed it back for OSX.
2. sounds like a minor issue could we bounty it ? The second part is same as in develop, it's because of the message status on the bottom right corner
3. fixed
4. it's in develop as well

**Comment by yenda** [comment]
5. regarding the slow down it's due to the fact that some stuff is now recalculated, I will work on optimizations later on. It's necessary otherwise you would see preview of blocked messages and counter would include blocked messages the way it was working before

**Comment by chadyj** [comment]
Bit more context for blocking >> https://github.com/status-im/status-react/issues/6735

**Comment by churik** [comment]
@yenda thank you!
### 2. First part is minor, agree, but second it is not working the same on current nightly and PR.

On current nightly:
in_outg

On PR build:
cur_pr
Difference is noticeable.

### 6. `Invalid arity:` 1 when creating account on IOS and Android
After creating account I've got an error:
android_png
After restarting app and login I see the same error.
Testing on mobile is blocked.

### 7. When you tap on "Block user" on profile, then click on "Send message", you  will be redirected to empty chat without UI

Video: http://take.ms/UvhXT
Log: [Status.log](https://github.com/status-im/status-react/files/2585002/Status.log)
I you tried to send some messages to this chat, receiver will receive them, but you cannot see them (only empty screen will be shown)
monosnap 2018-11-15 12-31-58


### 8. `TypeError: e.call is not a function` when tried to block a user who is not in my contact list
Steps:
- created new account on release build 0.9.30 (User A)
- login to desktop app (User B)
- click on "Share my contact code" (User B)
- added user B as contact (user A)
- block user  from his profile (User B)
error_0
Log: [Status.log](https://github.com/status-im/status-react/files/2585040/Status.log)

### 9. Messages that are coming from blocked user are counted in unread message badge on Mac OSx

This badge is updated when messages from blocked user are coming.
unread_0

### 10. Should be optimized in terms of performance before merge.
Tests were made on the same account with same chats on different builds.

Action | CPU, % (PR) | Real memory size (PR) | CPU, % (nightly) | Real memory size (nightly) | % increase on Memory Size in PR build
-- | -- | -- | -- | -- | --
Login | 1.79 | 429 | 3 | 385.9 | 11.17
Switching between chats (10 times) | 26.52 | 1010.9 | 27 | 512 | **97.44**
Join chat with 10+ messages | 3 | 591.2 | 11 | 469.2 | 26
Send 10 messages in a row | 1.6 | 772.9 | 1.79 | 500.3 | **54.49**
Switching to Profile and back (5 times) | 27 | 444.4 | 35 | 433.5 | 2.51
Scrolling chat history | 98 | 3190 | 27 | 704.9 | **352.55**
Switching mailserver | 9 | 665.3 | 4 | 626.4 | 6.21




<blockquote><div>Monosnap screenshot tool</div><div><strong><a href="https://monosnap.com/file/a8tGOVyO2eXILSvRPPrZXYIdFgKLUV">File "screencast 2018-11-15 12-08-58.mp4"</a></strong></div><div>Monosnap — the best tool to take & share your screenshots.</div></blockquote>

**Comment by yenda** [comment]
@churik 

I fixed 7 and 8, can you test performances again they should be better now

File: src/status_im/chat/db.cljs
Comment (rasom) [inline src/status_im/chat/db.cljs:44]: @yenda i believe we agreed once to put comments before the line instead of in the end of line

File: src/status_im/chat/db.cljs
Comment (rasom) [inline src/status_im/chat/db.cljs:53]: 80 chars limit

File: src/status_im/chat/db.cljs
Comment (rasom) [inline src/status_im/chat/db.cljs:18]: function is never used

File: src/status_im/chat/db.cljs
Comment (rasom) [inline src/status_im/chat/db.cljs:2]: namespace is never used

File: src/status_im/chat/db.cljs
Comment (rasom) [inline src/status_im/chat/db.cljs:134]: 80 chars, please check it in the whole ns, many places where it is exceeded

File: src/status_im/chat/db.cljs
Comment (rasom) [inline src/status_im/chat/db.cljs:147]: `response-to` is never used

File: src/status_im/chat/db.cljs
Comment (rasom) [inline src/status_im/chat/db.cljs:276]: `unviewed-messages` is never used, there are also other unused symbols in the ns, please check them

**Review by rasom — CHANGES_REQUESTED** [review]
Why is this PR that big? As far as i can see it was approved when its description was much smaller and scope was different. Can it be split into separate PRs? Should we request re-reviews from @flexsurfer  and @dmitryn?

**Comment by churik** [comment]
@yenda 
results from https://github.com/status-im/status-react/pull/6652#issuecomment-439285992

Action | CPU, % (PR) | Real memory size (PR) | CPU, % (nightly) | Real memory size (nightly) | % increase on Memory Size in PR build
-- | -- | -- | -- | -- | --
Login | 1.8 | 842 | 3 | 385.9 | **118.2**
Switching between chats (10 times) | 39 | 753.5 | 27 | 512 | 47.2
Join chat with 10+ messages | 1.3 | 598.9 | 11 | 469.2 | 27.6
Send 10 messages in a row | 16 | 734 | 1.79 | 500.3 | 46.7
Switching to Profile and back (5 times) | 34 | 490 | 35 | 433.5 | 13
Scrolling chat history | 91 | 2520 | 27 | 704.9 | **257.5**
Switching mailserver | 8 | 530 | 4 | 626.4 | -15.4


Also I noticed that message ordering is broken now, so if you send several message to channel, it is shown in the middle of the history or overlapping with other messages.

2_order
1_order

**Comment by yenda** [comment]
closed as refactoring was moved to https://github.com/status-im/status-react/pull/6764 and block feature will be in a separate PR

**Pull request closed without merging**
